src/types/pick.ts

```diff
@@ -0,0 +1,103 @@
+import type {
+    BasicType,
+    MessageDetails,
+    OneOrMore,
+    PossibleDiscriminator,
+    Properties,
+    PropertiesInfo,
+    Result,
+    Type,
+    TypeImpl,
+    TypeOf,
+    ValidationOptions,
+    Visitor,
+} from '../interfaces.js';
+import { decodeOptionalName } from '../utils/collection-utils.js';
+import { define } from '../utils/define.js';
+import { unknownRecord } from './unknown.js';
+import { hasOwnProperty } from '../utils/type-utils.js';
+import { prependPathToDetails } from '../utils/failure-utils.js';
+import { interfaceStringify } from '../utils/stringifiers.js';
+import { propsInfoToProps } from './union.js';
+
+export class PickType<Type extends BaseObjectLikeTypeImpl<unknown>, ResultType> extends BaseObjectLikeTypeImpl<ResultType> {
+    readonly name: string;
+    readonly isDefaultName: boolean;
+    readonly props: Properties;
+    override propsInfo: PropertiesInfo;
+    constructor(
+        readonly type: Type,
+        readonly keys: OneOrMore<keyof TypeOf<Type>>,
+        name?: string,
+    ) {
+        super();
+        this.isDefaultName = !name;
+        this.name = name || `Pick<${type.name}, ${keys.map(k => `'${k.toString()}'`).join(' | ')}>`;
+        const stringKeys = this.keys.map(k => k.toString()) as [string, ...string[]];
+        this.propsInfo = pickPropertiesInfo(this.type.propsInfo, stringKeys);
+        this.props = propsInfoToProps(this.propsInfo);
+        this.possibleDiscriminators = this.type.possibleDiscriminators;
+    }
+    override possibleDiscriminators: readonly PossibleDiscriminator[] = this.type.possibleDiscriminators; // TODO: Filter
+    override basicType!: BasicType;
+    override typeConfig: unknown;
+
+    protected override typeValidator(input: unknown, options: ValidationOptions): Result<ResultType> {
+        if (!unknownRecord.is(input)) {
+            return this.createResult(input, undefined, { kind: 'invalid basic type', expected: 'object' });
+        }
+
+        const constructResult: Record<string, unknown> = {};
+        const details: MessageDetails[] = [];
+        for (const [key, innerType] of this.propsArray) {
+            const missingKey = !hasOwnProperty(input, key);
+            const partialKey = this.propsInfo[key]?.partial;
+
+            if (missingKey) {
+                partialKey || details.push({ kind: 'missing property', property: key, type: innerType });
+                continue;
+            }
+
+            const innerResult = innerType.validate(input[key], options);
+            if (innerResult.ok) {
+                constructResult[key] = innerResult.value;
+            } else {
+                details.push(...prependPathToDetails(innerResult, key));
+            }
+        }
+        return this.createResult(input, options.mode === 'construct' ? constructResult : input, details);
+    }
+    override accept<R>(visitor: Visitor<R>): R {
+        // TODO: Should be fine, right?
```

Comment:
Yeah, once you distribute over unions in the `pick` function, then all that remains here is object-like types.